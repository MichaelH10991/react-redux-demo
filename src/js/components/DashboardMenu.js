import React, { useState, useMemo, useEffect, useCallback } from 'react'
import {
    Toolbar,
    AppBar,
    Button,
    Typography,
    Dialog,
    DialogContent,
    Menu,
    MenuItem,
    useTheme,
    Divider,
} from '@material-ui/core'
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown'
import DASHBOARDS from './data'
import { Link } from 'react-router-dom'

const DROPDOWN_ICON = <ArrowDropDownIcon />
const MENU_POSITION_PROPS = {
    anchorOrigin: { horizontal: 'left', vertical: 'bottom' },
    transformOrigin: { horizontal: 'left', vertical: 'top' },
}

function useLocation() {
    const [location, setLocation] = useState(window.location)
    useEffect(() => {
        const handler = () => setLocation(window.location)
        window.addEventListener('popstate', handler)
        return () => {
            window.removeEventListener('popstate', handler)
        }
    }, [])
    return location
}

function useMenu(initialState = null) {
    const [menuAnchor, setMenuAnchor] = useState(initialState)
    const handleOpenClick = useCallback(
        clickEvent => setMenuAnchor(clickEvent.currentTarget),
        [setMenuAnchor]
    )
    const handleClose = useCallback(() => setMenuAnchor(null), [])
    return { menuAnchor, handleOpenClick, handleClose }
}

function DashboardMenu() {
    // from useMenu() hook from ./MenuButton
    const { menuAnchor, handleOpenClick, handleClose } = useMenu()
    const pathname = useLocation().pathname

    const BUTTON_STYLE = {
        background: 'red',
        color: 'white',
    }
    return (
        <>
            <Button
                style={BUTTON_STYLE}
                onClick={handleOpenClick}
                endIcon={DROPDOWN_ICON}
            >
                Dashboards
            </Button>
            <Menu
                open={Boolean(menuAnchor)}
                anchorEl={menuAnchor}
                onClose={handleClose}
                onClick={handleClose}
                getContentAnchorEl={null}
                {...MENU_POSITION_PROPS}
            >
                {Object.entries(DASHBOARDS).map(([dashboardId, dashboard]) => {
                    const {
                        title,
                        url = `/dashboard/${dashboardId}`,
                    } = dashboard
                    const isCurrent = pathname.endsWith(url)
                    if (isCurrent) {
                        return (
                            <MenuItem key={dashboardId}>
                                <strong>{title}</strong>
                            </MenuItem>
                        )
                    }
                    return (
                        <MenuItem key={dashboardId} component={Link} to={url}>
                            {title}
                        </MenuItem>
                    )
                })}
                {
                    <>
                        <Divider />
                        <p>custom_dashboard</p>
                    </>
                }
            </Menu>
        </>
    )
}

export default DashboardMenu
