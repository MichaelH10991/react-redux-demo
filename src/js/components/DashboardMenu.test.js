import React from 'react'
import DashboardMenu from './DashboardMenu'
import renderer from 'react-test-renderer'

test('menu opens when clicked', () => {
    const component = renderer.create(<DashboardMenu />)
    // let tree = component.toJSON()
    expect(component).toMatchSnapshot()

    component.proto.onClick()
    expect(component).toMatchSnapshot()
})
