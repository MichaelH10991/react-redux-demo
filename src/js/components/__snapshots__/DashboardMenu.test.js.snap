// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu opens when clicked 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "background": "red",
      "color": "white",
    }
  }
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Dashboards
    <span
      className="MuiButton-endIcon MuiButton-iconSizeMedium"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </span>
  </span>
</button>
`;
